File: src/main/java/io/github/starwishsama/namelessbot/commands/GuessNumberCommand.java
package io.github.starwishsama.namelessbot.commands;

import cc.moecraft.icq.command.CommandProperties;
import cc.moecraft.icq.command.interfaces.GroupCommand;
import cc.moecraft.icq.event.events.message.EventGroupMessage;
import cc.moecraft.icq.user.Group;
import cc.moecraft.icq.user.GroupUser;
import cc.moecraft.utils.StringUtils;
import cn.hutool.core.util.RandomUtil;
import io.github.starwishsama.namelessbot.BotConstants;
import io.github.starwishsama.namelessbot.BotMain;
import io.github.starwishsama.namelessbot.managers.SessionManager;
import io.github.starwishsama.namelessbot.session.commands.guessnumber.GuessNumberSession;
import io.github.starwishsama.namelessbot.utils.BotUtils;

import java.util.ArrayList;
import java.util.Random;

/**
 * @author Nameless
 */
public class GuessNumberCommand implements GroupCommand {
    @Override
    public String groupMessage(EventGroupMessage event, GroupUser sender, Group group, String command, ArrayList<String> args) {
        if (BotUtils.isNoCoolDown(sender.getId())) {
            if (SessionManager.getSessionByGroup(group.getId()) != null){
                return BotUtils.getLocalMessage("msg.bot-prefix") + "你还没猜出数字呢!";
            } else {
                if (args.isEmpty()) {
                    event.respond(BotUtils.getLocalMessage("msg.bot-prefix") + "正在生成数字...");
                    GuessNumberSession session = new GuessNumberSession(sender.getId(), event.getGroupId(), new Random().nextInt(BotConstants.cfg.getMaxNumber()));
                    SessionManager.addSession(session);
                    BotMain.getLogger().log("[猜数字] 已生成随机数: " + session.getAnswer() + " 群聊: " + group.getId());
                    return BotUtils.getLocalMessage("msg.bot-prefix") + "请猜一个数字 (范围: [0," + BotConstants.cfg.getMaxNumber() + "])";
                } else {
                    if (args.size() > 1) {
                        if (StringUtils.isNumeric(args.get(0)) && StringUtils.isNumeric(args.get(1))) {
                            double min = Double.parseDouble(args.get(0));
                            double max = Double.parseDouble(args.get(1));
                            if (args.get(0).contains(".") || args.get(1).contains(".")) {
                                return BotUtils.getLocalMessage("msg.bot-prefix") + "不支持小数";
                            } else if (min == max){
                                return BotUtils.sendLocalMessage("msg.bot-prefix" + "两个数字不能相等!");
                            } else {
                                GuessNumberSession session = new GuessNumberSession(sender.getId(), event.getGroupId(), RandomUtil.randomInt(Integer.parseInt(args.get(0)), Integer.parseInt(args.get(1))));
                                SessionManager.addSession(session);
                                BotMain.getLogger().log("[猜数字] 已生成随机数: " + session.getAnswer() + " 群聊: " + group.getId());
                                return BotUtils.getLocalMessage("msg.bot-prefix") + "请猜一个数字 (范围: [" + args.get(0) + "," + args.get(1) + "])";
                            }
                        } else {
                            return BotUtils.getLocalMessage("msg.bot-prefix") + "请输入有效数字!";
                        }
                    }
                }
            }
        }
        return null;
    }

    @Override
    public CommandProperties properties() {
        return new CommandProperties("guess", "猜数字");
    }
}

File: src/main/kotlin/io/github/starwishsama/comet/commands/subcommands/chats/GuessNumberCommand.kt
package io.github.starwishsama.comet.commands.subcommands.chats

import cn.hutool.core.util.RandomUtil
import io.github.starwishsama.comet.BotVariables
import io.github.starwishsama.comet.commands.CommandProps
import io.github.starwishsama.comet.commands.interfaces.ChatCommand
import io.github.starwishsama.comet.commands.interfaces.SuspendCommand
import io.github.starwishsama.comet.enums.UserLevel
import io.github.starwishsama.comet.objects.BotUser
import io.github.starwishsama.comet.sessions.Session
import io.github.starwishsama.comet.sessions.SessionManager
import io.github.starwishsama.comet.sessions.commands.guessnumber.GuessNumberSession
import io.github.starwishsama.comet.sessions.commands.guessnumber.GuessNumberUser
import io.github.starwishsama.comet.utils.BotUtil
import io.github.starwishsama.comet.utils.isNumeric
import io.github.starwishsama.comet.utils.toMsgChain
import net.mamoe.mirai.contact.nameCardOrNick
import net.mamoe.mirai.message.GroupMessageEvent
import net.mamoe.mirai.message.MessageEvent
import net.mamoe.mirai.message.data.EmptyMessageChain
import net.mamoe.mirai.message.data.MessageChain
import net.mamoe.mirai.message.data.content
import java.time.Duration
import java.time.LocalDateTime

class GuessNumberCommand : ChatCommand, SuspendCommand {
    override suspend fun execute(event: MessageEvent, args: List<String>, user: BotUser): MessageChain {
        if (BotUtil.isNoCoolDown(user.id) && event is GroupMessageEvent) {
            val session = SessionManager.getSessionByGroup(event.group.id)
            if (session == null) {
                when {
                    args.isEmpty() -> {
                        val answer = RandomUtil.randomInt(0, 100)
                        BotVariables.logger.verbose("[猜数字] 群 ${event.group.id} 生成的随机数为 $answer")
                        SessionManager.addSession(GuessNumberSession(event.group.id, RandomUtil.randomInt(0, 101)))
                        return BotUtil.sendMessage("来猜个数字吧! 范围 [0, 100]")
                    }
                    args.size == 2 -> {
                        try {
                            val min = args[0].toInt()
                            val max = args[1].toInt()
                            if (min <= 0 || max <= 0) {
                                throw NumberFormatException("不支持负数")
                            }

                            if (min >= max) {
                                throw NumberFormatException("最小值不能大于等于最大值")
                            }
                            val answer = RandomUtil.randomInt(min, max + 1)
                            BotVariables.logger.verbose("[猜数字] 群 ${event.group.id} 生成的随机数为 $answer")
                            SessionManager.addSession(GuessNumberSession(event.group.id, RandomUtil.randomInt(0, 100)))
                            return BotUtil.sendMessage("猜一个数字吧! 范围 [$min, $max]")
                        } catch (e: NumberFormatException) {
                            return BotUtil.sendMessage("${e.message}")
                        }
                    }
                    else -> {
                        return getHelp().toMsgChain()
                    }
                }
            } else {
                return BotUtil.sendMessage("已经有一个游戏在进行中啦~")
            }
        }
        return EmptyMessageChain
    }

    override fun getProps(): CommandProps
            = CommandProps("guessnumber", arrayListOf("csz"), "猜数字", "nbot.commands.guessnumber", UserLevel.USER)

    override fun getHelp(): String = """
        /csz 猜数字
        /csz [最小值] [最大值] 猜指定范围内的数字
    """.trimIndent()

    override suspend fun handleInput(event: MessageEvent, user: BotUser, session: Session) {
        val trueAnswer = (session as GuessNumberSession).answer
        session.lastAnswerTime = LocalDateTime.now()
        val answer = event.message.content
        var gnUser = session.getGNUser(user.id)

        if (answer.isNumeric()) {
            val answerInInt = answer.toInt()

            if (gnUser == null) {
                gnUser = GuessNumberUser(event.sender.id, event.sender.nameCardOrNick)
                session.users += gnUser
            }

            gnUser.guessTime += 1

            when {
                answerInInt > trueAnswer -> {
                    event.reply(BotUtil.sendMessage("你猜的数字大了").contentToString())
                }
                answerInInt < trueAnswer -> {
                    event.reply(BotUtil.sendMessageToString("你猜的数字小了"))
                }
                answerInInt == trueAnswer -> {
                    session.usedTime = Duration.between(session.startTime, LocalDateTime.now())
                    val sb = StringBuilder(BotUtil.sendMessageToString("${event.sender.nameCardOrNick} 猜对了!\n总用时: ${session.usedTime.seconds}s\n\n"))
                    val list = session.users.sortedBy { (it as GuessNumberUser).guessTime }
                    list.forEach {
                        sb.append((it as GuessNumberUser).username).append(" ").append(it.guessTime).append("次\n")
                    }
                    event.reply(sb.toString().trim())
                    SessionManager.expireSession(session)
                }
            }
        } else {
            when (answer) {
                "不玩了", "结束游戏", "退出游戏" -> {
                    SessionManager.expireSession(session)
                    event.reply(BotUtil.sendMessageToString("游戏已结束"))
                }
            }
        }
    }
}